src/SAML2/Utils.php

```diff
@@ -649,23 +650,30 @@ public static function addStrings(
     /**
      * Create a KeyDescriptor with the given certificate.
      *
-     * @param string $x509Data The certificate, as a base64-encoded DER data.
+     * @param string|null $x509Data The certificate, as a base64-encoded DER data.
+     * @param string|null $keyName The name of the key as specified in the KeyInfo
      * @return \SAML2\XML\md\KeyDescriptor The keydescriptor.
      */
-    public static function createKeyDescriptor(string $x509Data) : KeyDescriptor
+    public static function createKeyDescriptor(string $x509Data = null, string $keyName = null) : KeyDescriptor
     {
-        $x509Certificate = new X509Certificate();
-        $x509Certificate->setCertificate($x509Data);
+        $keyInfo = new KeyInfo();
 
-        $x509Data = new X509Data();
-        $x509Data->addData($x509Certificate);
+        if ($keyName !== null) {
+            $keynameEl = new KeyName();
+            $keynameEl->setName($keyName);
+            $keyInfo->addInfo($keynameEl);
+        }
 
-        $keyInfo = new KeyInfo();
-        $keyInfo->addInfo($x509Data);
+        if ($x509Data !== null) {
+            $x509Certificate = new X509Certificate();
+            $x509Certificate->setCertificate($x509Data);
+            $x509Data = new X509Data();
+            $x509Data->addData($x509Certificate);
+            $keyInfo->addInfo($x509Data);
+        }
 
         $keyDescriptor = new KeyDescriptor();
         $keyDescriptor->setKeyInfo($keyInfo);
```

Comment:
The old code expected a value in `$x509Data`. However, the new implementation allows for both `$x509Data` and `$keyInfo` to be missing, and that would imply an empty `KeyInfo` element being added here. According to the schema, `KeyInfo` must have at least one child element, so our code could potentially allow for building non schema-compliant SAML documents.

I guess the new implementation should enforce that one value or the other are present, but never both absent at the same time.

Reply:
I agree. I added a check and a throw.